src/ossl/aes.rs

```diff
@@ -1654,45 +1713,53 @@ impl Encryption for AesOperation {
     /// May return different values depending on the internal status and the
     /// mode of operation.
     fn encryption_len(&mut self, data_len: usize, fin: bool) -> Result<usize> {
-        if self.finalized {
-            return Err(CKR_OPERATION_NOT_INITIALIZED)?;
-        }
         let outlen = if fin {
-            if !self.in_use {
-                return Err(CKR_OPERATION_NOT_INITIALIZED)?;
-            }
             match self.mech {
-                CKM_AES_CCM | CKM_AES_GCM => {
-                    if self.finalbuf.len() != 0 {
+                CKM_AES_CCM => {
+                    if data_len + self.finalbuf.len() > self.params.datalen {
                         return Err(self.op_err(CKR_DATA_LEN_RANGE));
                     }
-                    self.params.taglen
+                    self.finalbuf.len() + data_len + self.params.taglen
+                }
+                CKM_AES_GCM => data_len + self.params.taglen,
+                CKM_AES_CTR | CKM_AES_CTS => data_len,
+                CKM_AES_CBC | CKM_AES_ECB => {
+                    if (self.finalbuf.len() + data_len) % AES_BLOCK_SIZE != 0 {
+                        return Err(self.op_err(CKR_DATA_LEN_RANGE));
+                    }
+                    self.finalbuf.len() + data_len
                 }
-                CKM_AES_CTR | CKM_AES_CTS | CKM_AES_CBC | CKM_AES_ECB => 0,
-                #[cfg(not(feature = "fips"))]
-                CKM_AES_CFB8 | CKM_AES_CFB1 | CKM_AES_CFB128 | CKM_AES_OFB => 0,
                 CKM_AES_CBC_PAD => {
-                    if self.finalbuf.len() > 0 {
-                        self.finalbuf.len()
+                    /*
+                     * The PKCS#7 padding adds always at least 1 byte, so it
+                     * can emit up to a full block of padding even if there
+                     * is no final block data to parse.
+                     */
+                    ((self.finalbuf.len() + data_len + AES_BLOCK_SIZE)
+                        / AES_BLOCK_SIZE)
+                        * AES_BLOCK_SIZE
+                }
+                CKM_AES_KEY_WRAP => {
+                    if data_len % 8 != 0 {
+                        return Err(self.op_err(CKR_DATA_LEN_RANGE));
                     } else {
-                        AES_BLOCK_SIZE
+                        data_len + 8
```

Comment:
just a though if we should make a constant out of the 8 here to be clear that this is the AES Key Wrap algorithm block size here and in the above check. And likely also in the KWP mechanism below.
In the decryption case there is already `(AES_BLOCK_SIZE / 2)` used, but having KWP define might make the code easier to read and understand.

Reply:
Added a commit with a new constant for this